### PR TITLE
Add page movement to zipf

### PR DESCRIPTION
**Pull request by maciekpac**
### Types of changes

- [ ] Bugfix (non-breaking change which fixes issue linked in Description above)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)
- [ ] Other

### Checklist

- [x] Code compiles without errors
- [x] All tests pass locally with my changes (see TESTING section in CONTRIBUTING file)
- [ ] Created tests which will fail without the change (if possible)
- [ ] Extended the README/documentation (if necessary)
- [ ] All newly added files have proprietary license (if necessary)
- [ ] All newly added files are referenced in MANIFEST files (if necessary)

## Further comments

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/769)

### CODE REVIEW

File: utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp [hunk 1]
```diff
@@ -539,6 +571,10 @@ create_loader_creators(TestCase test_case, size_t nof_allocations,
                     break;
                 case RawAllocator::FAST_POOL:
                     allocator = std::make_shared<FastPoolAllocatorWrapper>();
```
Comment (PatKamin) [inline utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp:573]: Add a missing `break`.

File: utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp [hunk 2]
```diff
@@ -696,6 +733,27 @@ static int parse_opt(int key, char *arg, struct argp_state *state)
...
+            assert(arg && "argument for allocation size not supplied");
+            args->limits.softLimit = std::strtoul(arg, nullptr, 10);
+            assert(args->limits.softLimit > 0 && "limit has to be > 0");
+            break;
+        }
+        case KEY_HARD_LIMIT:
+        {
+            assert(arg && "argument for allocation size not supplied");
+            args->limits.hardLimit = std::strtoul(arg, nullptr, 10);
+            assert(args->limits.hardLimit > 0 && "limit has to be > 0");
```
Comment (PatKamin) [inline utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp:754]: Check if `hard_limit` is greater than `soft_limit`.

**Review by PatKamin — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/769)* status: all files reviewed, 3 unresolved discussions (waiting on @bratpiorka and @maciekpac)

___
*[utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp, line 575 at r1](https://reviewable.io/reviews/memkind/memkind/769#-MyHb50dmgrq8G-BY5pz-r1-575:-MyIEE0G6XFU4PGMxV_0:bp1pmwv) ([raw file](https://github.com/memkind/memkind/blob/211127cdd8e841e36170179d5866c5b0d94f8e1b/utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp#L575)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

AFAIK in this case limits->lowLimit would not be checked. But shouldn't this if be negated with '!' ?:

if !(limits && limits->lowLimit && limits->softLimit && limits->hardLimit) {..}
</blockquote></details>

Indeed, the if statement should be negated, thank you. The assert could be checked with a negative test to make sure what is the behavior when some nullptrs are passed.

**Review by maciekpac — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/769)* status: all files reviewed, 3 unresolved discussions (waiting on @PatKamin)

___
*[utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp, line 573 at r1](https://reviewable.io/reviews/memkind/memkind/769#-MyHb50dmgrq8G-BY5pz-r1-573:-MyIZIqx8hx6rp0wdACa:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/211127cdd8e841e36170179d5866c5b0d94f8e1b/utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp#L573)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Add a missing `break`.
</blockquote></details>

Done.
___
*[utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp, line 575 at r1](https://reviewable.io/reviews/memkind/memkind/769#-MyHb50dmgrq8G-BY5pz-r1-575:-MyIZ_Ru3_CQ-gV24OVQ:be6sz35) ([raw file](https://github.com/memkind/memkind/blob/211127cdd8e841e36170179d5866c5b0d94f8e1b/utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp#L575)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Indeed, the if statement should be negated, thank you. The assert could be checked with a negative test to make sure what is the behavior when some nullptrs are passed.
</blockquote></details>

Why would limits->lowLimit be checked, if limits is nullptr ?

in C, if left handside of && operator evaluates to false, to right handside is not checked:

https://stackoverflow.com/questions/5683026/logical-and-or-is-left-to-right-evaluation-guaranteed

please correct me if I'm messing something up
___
*[utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp, line 754 at r1](https://reviewable.io/reviews/memkind/memkind/769#-MyHb50dmgrq8G-BY5pz-r1-754:-MyI_V7-ByZDTv2UvW16:b7jeraq) ([raw file](https://github.com/memkind/memkind/blob/211127cdd8e841e36170179d5866c5b0d94f8e1b/utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp#L754)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Check if `hard_limit` is greater than `soft_limit`.
</blockquote></details>

any such checks are performed internally, in mtt_allocators (at least should be); I'm checking for "0" because 0 is used to mark the value as "uninitialized"

**Review by PatKamin — CHANGES_REQUESTED** [review]
> Reviewed 4 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/769)* status: 4 of 5 files reviewed, 1 unresolved discussion (waiting on @bratpiorka and @maciekpac)

___
*[utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp, line 575 at r1](https://reviewable.io/reviews/memkind/memkind/769#-MyHb50dmgrq8G-BY5pz-r1-575:-MyLYCr_6o_uXq5e3p5i:bqabo1h) ([raw file](https://github.com/memkind/memkind/blob/211127cdd8e841e36170179d5866c5b0d94f8e1b/utils/memtier_zipf_bench/memtier_zipf_counter_bench.cpp#L575)):*
<details><summary><i>Previously, maciekpac wrote…</i></summary><blockquote>

Why would limits->lowLimit be checked, if limits is nullptr ?

in C, if left handside of && operator evaluates to false, to right handside is not checked:

https://stackoverflow.com/questions/5683026/logical-and-or-is-left-to-right-evaluation-guaranteed

please correct me if I'm messing something up
</blockquote></details>

You're right, I've misinterpreted what's going on here:
https://github.com/memkind/memkind/blob/master/src/memkind_log.c#L92-L93

**Review by PatKamin — APPROVED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/769)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @maciekpac)

**Review by bratpiorka — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/769)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @maciekpac)

**Pull request merged**